update-readme: add single make target (#453)
lambdaclass/ethereum_rust: README.md

2. Prioritize clarity
3. Embrace simplicity in design

We believe this approach is the best way to build a client that is both fast and resilient. By adhering to these principles, we will be able to iterate fast and explore next-generation features early, either from the Ethereum roadmap or from innovations from the L2s. 

## Quick Start (localnet)
![Demo][./demo.png]
You'll need docker, docker compose and rust 1.81.
### Mac
```shell
brew install kurtosis-tech/tap/kurtosis-cli && make
```
### Linux
[Install kurtosis][https://docs.kurtosis.com/install/#ii-install-the-cli] and run:
```shell
make
```

This will:
- Start [ethereum package][https://github.com/ethpandaops/ethereum-package], a private testnet on which multiple ethereum clients can interact.
- Start our execution client, and make it interact with the others.

## Dev Setup

> ### Build

Reviewer: Mixed feelings about this, I would recomment asdf instead since we don't necesarily support the latest rust version (although we probably should).

We believe this approach is the best way to build a client that is both fast and resilient. By adhering to these principles, we will be able to iterate fast and explore next-generation features early, either from the Ethereum roadmap or from innovations from the L2s.

## Quick Start (localnet)

> ![Demo](https://raw.githubusercontent.com/lambdaclass/ethereum_rust/update-readme/Demo.png)

Reviewer: you have a link to the branch here.

Author: Should be a permalink now.

> You'll need docker, docker compose, clang and rust 1.80

Reviewer: you sure we need all these? I wouldn't reference a particular rust version because otherwise we have to update it every time.

Author: You're right about not specifying a rust version. I'm sure we need Docker, not so sure about docker-compose, I think some dependency fails without clang, but I'll double check it and let you know.

> ### Mac

Reviewer: I would just use the link that you added in the linux part fo everything, since it has instructions for every os.

Author: I've unified everything into a single 'Setup' header, wdy?

### Prerequisites
- [Kurtosis](https://docs.kurtosis.com/install/#ii-install-the-cli)
- Rust.
- Docker.
```shell
make localnet
```

This make target will:
1. Build our node inside a docker image.
2. Fetch our fork [ethereum package](https://github.com/ethpandaops/ethereum-package), a private testnet on which multiple ethereum clients can interact.
3. Start the localnet with kurtosis.

If everything went well, you should be faced with our client's logs (ctrl-c to leave)

To stop everything, simply run:
```shell
make stop-localnet
```

## Dev Setup
### Build

#### Rust
To build the node, you will need the rust toolchain: 
1. First, [install asdf](https://asdf-vm.com/guide/getting-started.html): 
2. Add the rust plugin: 
```shell
asdf plugin-add rust https://github.com/asdf-community/asdf-rust.git
```
3. cd into the project and run:
```shell
asdf install
```

You now should be able to build the client:
```bash
make build
```

### Testing

For testing, we're using three kinds of tests.

##### Ethereum Foundation Tests

These are the official execution spec tests, tests that should be implementation agnostic, you can execute them with:

```bash

> make test

Reviewer 2: ```suggestion
make test ef-tests
```
test runs all of the tests

```

> This will previously download it from the [official execution spec tests repo](https://github.com/ethereum/execution-spec-tests/) and run them with our glue code

Reviewer 2: ```suggestion
This will download the test cases from the [official execution spec tests repo](https://github.com/ethereum/execution-spec-tests/) and run them with our glue code
```
nit

under `cmd/ef_tests/tests/cancun.rs`.

##### Crate Specific Tests

> The second kind are each crate tests, you can run it like this:

Reviewer 2: ```suggestion
The second kind are each crate's tests, you can run them like this:
```
nit: language

```bash
make test CRATE=<crate>
```

##### Hive Tests

Finally, we have End-to-End tests with hive.
Hive is a system which simply sends RPC commands to our node,
and expects a certain response. You can read more about it [here](https://github.com/ethereum/hive/blob/master/docs/overview.md).
And you can run this set of tests with:

```bash
make run-hive-debug ethereum/rpc-compat
```

### Localnet

> To run a localnet, we'll use our fork of [ethPandaOps' Ethereum Package](https://github.com/ethpandaops/ethereum-package), specifically [this branch](https://github.com/lambdaclass/ethereum-package/tree/ethereum-rust-integration) that adds support to our client. We have that included in our repo as a `make` target. Make sure to fetch it like follows:

Reviewer 2: ```suggestion
To run a localnet, we'll use our fork of [ethPandaOps' Ethereum Package](https://github.com/ethpandaops/ethereum-package), specifically [this branch](https://github.com/lambdaclass/ethereum-package/tree/ethereum-rust-integration) that adds support to for client. We have that included in our repo as a `make` target. Make sure to fetch it as follows:
```
nit: language

```bash
make checkout-ethereum-package
```

If you haven't yet, install Kurtosis:

```bash
# Make sure to have docker installed

> brew install docker

Reviewer 2: Maybe we should link the official docker installation page instead so that users with different systems can also use it? https://docs.docker.com/engine/install/

# Kurtosis cli

> brew install kurtosis-tech/tap/kurtosis-cli

Reviewer 2: Same here: https://docs.kurtosis.com/install/

make localnet
```

Here, we setup kurtosis and make it run with our node inside a docker file.

> The config in charge of doing this is the file in under `/test_data/network_params.yaml`

Reviewer 2: ```suggestion
The config in charge of doing this is under the file `/test_data/network_params.yaml`
```
